pkg/spdk/engine.go

```diff
@@ -1229,6 +1351,396 @@ func (e *Engine) validateAndUpdateReplicaNvme(replicaName string, bdev *spdktype
+			return suspendFrontend, bdevUUID, err
+		}
+	} else if !deleted {
+		return suspendFrontend, bdevUUID, fmt.Errorf("engine %s raid delete failed", e.Name)
+	}
+
+	return suspendFrontend, bdevUUID, nil
+}
+
+func (e *Engine) expandReplicas(replicaClients map[string]*client.SPDKClient, spdkClient *spdkclient.Client, size uint64) (err error) {
+	e.log.Info("Expand replicas")
+
+	var (
+		errorLock sync.Mutex
+		wg        sync.WaitGroup
+	)
+
+	failedReplica := map[string]error{}
+
+	for replicaName, replicaClient := range replicaClients {
+		wg.Add(1)
+
+		go func(replicaName string, replicaClient *client.SPDKClient) {
+			defer wg.Done()
+			defer func() {
+				if r := recover(); r != nil {
+					e.log.WithField("replica", replicaName).Errorf("Panic during replica expansion: %v", r)
+				}
+			}()
+
+			replicaStatus, ok := e.ReplicaStatusMap[replicaName]
+			if !ok {
+				e.log.WithField("replica", replicaName).Warn("Replica not found in status map")
+				return
+			}
+
+			replica, err := replicaClient.ReplicaGet(replicaName)
+			if err != nil {
+				errorLock.Lock()
+				failedReplica[replicaName] = errors.Wrap(err, "Get replica failure")
+				errorLock.Unlock()
+				return
+			}
+
+			if replica.SpecSize == size {
+				return
+			}
+
+			if err := disconnectNVMfBdev(spdkClient, replicaStatus.BdevName); err != nil {
+				errorLock.Lock()
+				failedReplica[replicaName] = err
+				errorLock.Unlock()
+				return
+			}
+
+			if err := replicaClient.ReplicaExpand(replicaName, size); err != nil {
+				errorLock.Lock()
+				failedReplica[replicaName] = err
+				errorLock.Unlock()
+				return
+			}
+
+		}(replicaName, replicaClient)
+	}
+
+	wg.Wait()
+
+	// Some replicas may have returned an error during expansion due to unexpected issues
+	// (e.g. temporary network glitch, internal error, timeout).
+	// To avoid mistakenly marking those as failed, we perform a single follow-up check
+	// to verify if the replica was actually expanded.
+	//
+	// If ReplicaGet shows the desired size, we consider the expansion successful
+	// and remove it from the failed list.
+	for replicaName := range failedReplica {
+		client := replicaClients[replicaName]
+
+		replica, err := client.ReplicaGet(replicaName)
+		if err != nil {
+			e.log.WithError(err).WithField("replica", replicaName).
+				Warn("Could not verify replica expansion after failure")
+			continue
+		}
+
+		if replica.SpecSize == size {
+			e.log.WithField("replica", replicaName).
+				Info("Replica expansion succeeded despite earlier error, removing from failure list")
+
+			delete(failedReplica, replicaName)
+		}
+	}
+
+	var aggregatedErr map[string]string
+
+	if len(failedReplica) > 0 {
+		aggregatedErr = map[string]string{}
+		for replicaName, err := range failedReplica {
+			aggregatedErr[replicaName] = err.Error()
+		}
+	}
+
+	switch {
+	case len(failedReplica) == 0:
+		// all success
+		e.log.Info("All replicas expand success")
+		return nil
+	case len(failedReplica) == len(e.ReplicaStatusMap):
```

Comment:
Shouldn’t this compare `len(failedReplica)` with `len(replicaClients)` instead? Could `e.ReplicaStatusMap` include replicas that weren’t part of this expansion and give the wrong result?